core/arch/riscv/kernel/thread_arch.c

```diff
@@ -0,0 +1,613 @@
+		.status = trap_regs->status,
+		.ra = trap_regs->ra,
+		.a0 = trap_regs->a0,
+		.a1 = trap_regs->a1,
+		.a2 = trap_regs->a2,
+		.a3 = trap_regs->a3,
+		.a4 = trap_regs->a4,
+		.a5 = trap_regs->a5,
+		.a6 = trap_regs->a6,
+		.a7 = trap_regs->a7,
+		.t0 = trap_regs->t0,
+		.t1 = trap_regs->t1,
+	};
+}
+
+static void thread_user_ecall_handler(struct thread_trap_regs *trap_regs)
+{
+	struct thread_scall_regs scall_regs;
+	struct thread_core_local *l = thread_get_core_local();
+	int ct = l->curr_thread;
+
+	copy_trap_to_scall(trap_regs, &scall_regs);
+	thread_scall_handler(&scall_regs);
+	copy_scall_to_trap(&scall_regs, trap_regs);
+	/*
+	 * Save kernel sp we'll had at the beginning of this function.
+	 * This is when this TA has called another TA because
+	 * __thread_enter_user_mode() also saves the stack pointer in this
+	 * field.
+	 */
+	threads[ct].kern_sp = (unsigned long)(trap_regs + 1);
+	/*
+	 * We are returning to U-Mode, on return, the program counter
+	 * is set to xsepc (pc=xepc), we add 4 (size of an instruction)
+	 * to continue to next instruction.
+	 */
+	trap_regs->epc += 4;
+}
+
+static void copy_trap_to_abort(struct thread_trap_regs *trap_regs,
+			       struct thread_abort_regs *abort_regs)
+{
+	*abort_regs = (struct thread_abort_regs) {
+		.status = trap_regs->status,
+		.ra = trap_regs->ra,
+		.sp = trap_regs->sp,
+		.gp = trap_regs->gp,
+		.tp = trap_regs->tp,
+		.t0 = trap_regs->t0,
+		.t1 = trap_regs->t1,
+		.t2 = trap_regs->t2,
+		.s0 = trap_regs->s0,
+		.s1 = trap_regs->s1,
+		.a0 = trap_regs->a0,
+		.a1 = trap_regs->a1,
+		.a2 = trap_regs->a2,
+		.a3 = trap_regs->a3,
+		.a4 = trap_regs->a4,
+		.a5 = trap_regs->a5,
+		.a6 = trap_regs->a6,
+		.a7 = trap_regs->a7,
+		.s2 = trap_regs->s2,
+		.s3 = trap_regs->s3,
+		.s4 = trap_regs->s4,
+		.s5 = trap_regs->s5,
+		.s6 = trap_regs->s6,
+		.s7 = trap_regs->s7,
+		.s8 = trap_regs->s8,
+		.s9 = trap_regs->s9,
+		.s10 = trap_regs->s10,
+		.s11 = trap_regs->s11,
+		.t3 = trap_regs->t3,
+		.t4 = trap_regs->t4,
+		.t5 = trap_regs->t5,
+		.t6 = trap_regs->t6,
+	};
+}
+
+static void thread_abort_handler(struct thread_trap_regs *trap_regs,
+				 unsigned long cause)
+{
+	struct thread_abort_regs abort_regs = { };
+
+	assert(cause == read_csr(CSR_XCAUSE));
+	copy_trap_to_abort(trap_regs, &abort_regs);
+	abort_regs.cause = read_csr(CSR_XCAUSE);
+	abort_regs.epc = read_csr(CSR_XEPC);
+	abort_regs.tval = read_csr(CSR_XTVAL);
+	abort_regs.satp = read_csr(CSR_SATP);
+	abort_handler(cause, &abort_regs);
+}
+
+static void thread_exception_handler(unsigned long cause,
+				     struct thread_trap_regs *regs)
+{
+	switch (cause) {
+	case CAUSE_USER_ECALL:
+		thread_user_ecall_handler(regs);
+		break;
+	default:
+		thread_abort_handler(regs, cause);
+		break;
+	}
+}
+
+static void thread_irq_handler(void)
+{
+	itr_core_handler();
+}
+
+static void thread_interrupt_handler(unsigned long cause,
+				     struct thread_trap_regs *regs)
+{
+	switch (cause) {
```

Comment:
The cause parameter needs to be processed,Otherwise, the system enters default .
for example: cause value is 0x8000000000000005;
it will not entry IRQ_XTIMER case.